src/Identifier/PasswordIdentifier.php

```diff
@@ -88,63 +88,61 @@ public function getPasswordHasher()
      */
     public function identify(array $data)
     {
-        $fields = $this->getConfig('fields');
-
-        $usernameFields = (array)$fields['username'];
-        $found = array_intersect_key(array_flip($usernameFields), $data);
-        if (empty($found)) {
+        if (!isset($data[self::CREDENTIAL_USERNAME])) {
             return null;
         }
 
-        $password = null;
-        if (!empty($data[$fields['password']])) {
-            $password = $data[$fields['password']];
+        $identity = $this->_findIdentity($data[self::CREDENTIAL_USERNAME]);
+        if (array_key_exists(self::CREDENTIAL_PASSWORD, $data)) {
+            $password = $data[self::CREDENTIAL_PASSWORD];
+            if (!$this->_checkPassword($identity, $password)) {
+                return null;
+            }
         }
 
-        return $this->_findIdentity($data[key($found)], $password);
+        return $identity;
     }
 
     /**
      * Find a user record using the username and password provided.
      * Input passwords will be hashed even when a user doesn't exist. This
      * helps mitigate timing attacks that are attempting to find valid usernames.
      *
-     * @param string $identifier The username/identifier.
-     * @param string|null $password The password, if not provided password checking is skipped
-     *   and result of find is returned.
-     * @return \ArrayAccess|null User data entity or null on failure.
+     * @param array|\ArrayAccess|null $identity The identity or null.
+     * @param string|null $password The password.
+     * @return bool
      */
-    protected function _findIdentity($identifier, $password = null)
+    protected function _checkPassword($identity, $password)
     {
-        $result = $this->_findUser($identifier);
-        if (empty($result)) {
-            return null;
-        }
+        $passwordField = $this->getConfig('fields.' . self::CREDENTIAL_PASSWORD);
 
-        if ($password !== null) {
-            $passwordField = $this->getConfig('fields.password');
-            $hasher = $this->getPasswordHasher();
-            $hashedPassword = $result[$passwordField];
-            if (!$hasher->check($password, $hashedPassword)) {
-                return null;
-            }
+        if ($identity === null) {
+            $identity = [
+                $passwordField => ''
+            ];
+        }
 
-            $this->_needsPasswordRehash = $hasher->needsRehash($hashedPassword);
-            unset($result[$passwordField]);
+        $hasher = $this->getPasswordHasher();
+        $hashedPassword = $identity[$passwordField];
+        if (!$hasher->check($password, $hashedPassword)) {
+            return false;
         }
 
-        return $result;
+        $this->_needsPasswordRehash = $hasher->needsRehash($hashedPassword);
+        unset($identity[$passwordField]);
```

Comment:
Why is this unset() necessary?

Reply:
@markstory I'm not sure. It's been there before. 

@burzum any idea?

Reply:
I guess to empty the password field in case of a form? No idea other than that. Just remove it and see if a test breaks. Is the rehashing tested btw?

Reply:
Some tests fail as they expect that the password field is missing. I'm still not sure why it's being unset. I'll remove this, OK?

Reply:
OK